go/services/adr/client.go

```diff
@@ -0,0 +1,539 @@
+// Licensed under the MIT License.
+
+package adr
+
+import (
+	"context"
+	"log/slog"
+	"sync"
+	"time"
+
+	"github.com/Azure/iot-operations-sdks/go/protocol"
+	"github.com/Azure/iot-operations-sdks/go/services/adr/adrnamegen/adrbaseservice"
+	"github.com/Azure/iot-operations-sdks/go/services/adr/adrtypegen/aeptypeservice"
+)
+
+const (
+	aepNameTokenKey = "aepName"
+	defaultTimeout  = 10 * time.Second
+)
+
+type (
+	Asset                          = adrbaseservice.Asset
+	AssetEndpointProfile           = adrbaseservice.AssetEndpointProfile
+	NotificationResponse           = adrbaseservice.NotificationResponse
+	AssetStatus                    = adrbaseservice.AssetStatus
+	AssetEndpointProfileStatus     = adrbaseservice.AssetEndpointProfileStatus
+	DetectedAsset                  = adrbaseservice.DetectedAsset
+	DiscoveredAssetEndpointProfile = aeptypeservice.DiscoveredAssetEndpointProfile
+)
+
+// Error represents an error returned by the ADR service.
+type Error struct {
+	Message       string
+	PropertyName  string
+	PropertyValue string
+	Condition     interface{}
+}
+
+func (e *Error) Error() string {
+	return e.Message
+}
+
+// Client manages interactions with the Azure Device Registry.
+type Client struct {
+	logger                                             *slog.Logger
+	protocol                                           *protocol.Application
+	mqtt                                               protocol.MqttClient
+	observedAeps                                       map[string]struct{}
+	observedAssets                                     map[string]struct{}
+	mu                                                 sync.RWMutex
+	closeOnce                                          sync.Once
+	listeners                                          protocol.Listeners
+	onAssetUpdate                                      func(aepName string, asset *Asset) error
+	onAepUpdate                                        func(aepName string, profile *AssetEndpointProfile) error
+	notifyOnAssetEndpointProfileUpdateInvoker          *adrbaseservice.NotifyOnAssetEndpointProfileUpdateCommandInvoker
+	notifyOnAssetUpdateInvoker                         *adrbaseservice.NotifyOnAssetUpdateCommandInvoker
+	getAssetEndpointProfileCommandInvoker              *adrbaseservice.GetAssetEndpointProfileCommandInvoker
+	updateAssetEndpointProfileStatusCommandInvoker     *adrbaseservice.UpdateAssetEndpointProfileStatusCommandInvoker
+	createDiscoveredAssetEndpointProfileCommandInvoker *aeptypeservice.CreateDiscoveredAssetEndpointProfileCommandInvoker
+	createDetectedAssetCommandInvoker                  *adrbaseservice.CreateDetectedAssetCommandInvoker
+	getAssetCommandInvoker                             *adrbaseservice.GetAssetCommandInvoker
+	updateAssetStatusCommandInvoker                    *adrbaseservice.UpdateAssetStatusCommandInvoker
+}
+
+// ClientOption allows setting optional parameters on the Client.
+type ClientOption func(*Client)
```

Comment:
nit: This is a bit different than how the rest of the SDK does options. It's the traditional functional options pattern, so I won't block on it, but figured I'd mention in case you care about the discrepancy.